File: additions/my_contribution/backend/app/schemas/report.py
from datetime import date, datetime

from pydantic import BaseModel

ReportType = str  # "daily" | "weekly" | "monthly" | "yearly" | "technician" | "client" | "project" | "contract"


class ReportTypeInfo(BaseModel):
    key: str
    label: str


class InterventionReportRow(BaseModel):
    bi_number: str
    intervention_date: date
    technician_name: str
    client_name: str
    site_name: str
    intervention_type: str
    status: str
    net_duration_minutes: int
    points_earned: int


class InterventionReportSummary(BaseModel):
    total_interventions: int
    total_duration_minutes: int
    total_points: int
    fully_approved_count: int
    rejected_count: int


class InterventionReport(BaseModel):
    report_type: str
    date_from: date | None
    date_to: date | None
    summary: InterventionReportSummary
    rows: list[InterventionReportRow]


class ApprovalReportRow(BaseModel):
    bi_number: str
    approval_level: str
    approver_name: str
    decision: str
    comment: str | None
    approval_date: datetime


class ApprovalReport(BaseModel):
    date_from: date | None
    date_to: date | None
    total_approved: int
    total_rejected: int
    rows: list[ApprovalReportRow]


class PlanningReportRow(BaseModel):
    technician_name: str
    client_name: str
    site_name: str
    planned_date: date
    priority: str
    status: str


class PlanningReport(BaseModel):
    date_from: date | None
    date_to: date | None
    total_planned: int
    total_cancelled: int
    rows: list[PlanningReportRow]


class ComparisonPeriodResult(BaseModel):
    label: str
    total_interventions: int
    total_duration_minutes: int
    total_points: int
    fully_approved_count: int
    rejected_count: int


class ComparisonReport(BaseModel):
    period_a: ComparisonPeriodResult
    period_b: ComparisonPeriodResult

File: additions/my_contribution/frontend/src/types/report.ts
export interface ReportTypeInfo {
  key: string;
  label: string;
}

export interface InterventionReportRow {
  bi_number: string;
  intervention_date: string;
  technician_name: string;
  client_name: string;
  site_name: string;
  intervention_type: string;
  status: string;
  net_duration_minutes: number;
  points_earned: number;
}

export interface InterventionReportSummary {
  total_interventions: number;
  total_duration_minutes: number;
  total_points: number;
  fully_approved_count: number;
  rejected_count: number;
}

export interface InterventionReport {
  report_type: string;
  date_from: string | null;
  date_to: string | null;
  summary: InterventionReportSummary;
  rows: InterventionReportRow[];
}

export interface ApprovalReportRow {
  bi_number: string;
  approval_level: string;
  approver_name: string;
  decision: string;
  comment: string | null;
  approval_date: string;
}

export interface ApprovalReport {
  date_from: string | null;
  date_to: string | null;
  total_approved: number;
  total_rejected: number;
  rows: ApprovalReportRow[];
}

export interface PlanningReportRow {
  technician_name: string;
  client_name: string;
  site_name: string;
  planned_date: string;
  priority: string;
  status: string;
}

export interface PlanningReport {
  date_from: string | null;
  date_to: string | null;
  total_planned: number;
  total_cancelled: number;
  rows: PlanningReportRow[];
}

export interface ComparisonPeriodResult {
  label: string;
  total_interventions: number;
  total_duration_minutes: number;
  total_points: number;
  fully_approved_count: number;
  rejected_count: number;
}

export interface ComparisonReport {
  period_a: ComparisonPeriodResult;
  period_b: ComparisonPeriodResult;
}
